### PR TITLE
[WebSub] Change hub startup function signature

### PR DESCRIPTION
**Pull request by chamil321**
```ballerina
public function startHub(http:Listener hubServiceListener, 
                         HubConfiguration? hubConfiguration = ()) 
                                    returns WebSubHub|HubStartedUpError {
 }
```
Record definitions
```ballerina
public type HubConfiguration record {
   int leaseSeconds = 86400;
   SignatureMethod signatureMethod = SHA256; 
   RemotePublishConfig remotePublish?;
   boolean topicRegistrationRequired = true;
   string publicUrl?;
   http:ClientEndpointConfig clientConfig?;
};

public type RemotePublishConfig record {
   boolean enabled = false;
   RemotePublishMode mode = REMOTE_PUBLISHING_MODE_DIRECT;
};

```
https://docs.google.com/document/d/1yJ9GU7FxsA4CHWDheXS2OpE8VleeGKMaH8GrLVm4Vko/edit?usp=sharing

```ballerina
websub:startHub(new http:Listener(9090));
```

### CODE REVIEW

File: stdlib/websub/src/main/ballerina/websub/commons.bal [hunk 1]
```diff
@@ -87,6 +85,15 @@ public const PUBLISH_MODE_DIRECT = "PUBLISH_MODE_DIRECT";
 # needs to fetch the topic URL to identify the update content.
 public const PUBLISH_MODE_FETCH = "PUBLISH_MODE_FETCH";
 
+# The identifier to be used to identify the cryptographic hash Algorithm.
+public type SignatureMethod SHA1|SHA256;
+
+# The constant used to represent SHA-1 cryptographic hash Algorithm
+public const string SHA1 = "SHA1";
+
+# The constant used to represent SHA-256 cryptographic hash Algorithm
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/commons.bal]: shall we use lower case `a` in these three docs.

File: stdlib/websub/src/main/ballerina/websub/commons.bal [hunk 2]
```diff
@@ -440,63 +447,59 @@ public type SubscriptionChangeResponse record {
...
 #
-# + host - The hostname to start up the hub on
-# + port - The port to start up the hub on
 # + leaseSeconds - The default lease seconds value to honour if not specified in subscription requests
 # + signatureMethod - The signature method to use for authenticated content delivery (`SHA1`|`SHA256`)
-# + remotePublishingEnabled - Whether remote publishers should be allowed to publish to this hub (HTTP requests)
-# + remotePublishMode - If remote publishing is allowed, the mode to use, `direct` (default) - fat ping with
-#                          the notification payload specified or `fetch` - the hub fetches the topic URL
-#                          specified in the "publish" request to identify the payload
+# + remotePublish - The record representing configs related to remote publishing allowance
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/commons.bal]: `configs` --> `configuration`

File: stdlib/websub/src/main/ballerina/websub/commons.bal [hunk 3]
```diff
@@ -440,63 +447,59 @@ public type SubscriptionChangeResponse record {
...
 # + topicRegistrationRequired - Whether a topic needs to be registered at the hub prior to publishing/subscribing
 #                               to the topic
 # + publicUrl - The URL for the hub to be included in content delivery requests, defaults to
 #               `http(s)://localhost:{port}/websub/hub` if unspecified
-# + sslEnabled - Whether SSL needs to be enabled for the hub, enabled by default
-# + serviceSecureSocket - The SSL configuration for the hub service endpoint
-# + clientSecureSocket - The SSL configuration for the hub service for secure communication with remote HTTP
-#                        endpoints
+# + clientConfig - The configurations for the hub to communicate with remote HTTP endpoints
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/commons.bal]: `configurations` --> `configuration`

File: stdlib/websub/src/main/ballerina/websub/commons.bal [hunk 4]
```diff
@@ -440,63 +447,59 @@ public type SubscriptionChangeResponse record {
...
+#                          specified in the "publish" request to identify the payload
+public type RemotePublishConfig record {
+    boolean enabled = false;
+    RemotePublishMode mode = PUBLISH_MODE_DIRECT;
+};
+
+# Starts up the Ballerina Hub.
+#
+# + hubServiceListener - The `http:Listener` to which the hub service is attached
+# + hubConfiguration - The hub specific configurations
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/commons.bal]: `configurations` --> `configuration`

File: stdlib/websub/src/main/ballerina/websub/commons.bal [hunk 5]
```diff
@@ -440,63 +447,59 @@ public type SubscriptionChangeResponse record {
...
+
+# Record representing remote publishing allowance.
+#
+# + enabled - Whether remote publishers should be allowed to publish to this hub (HTTP requests)
+# + mode - If remote publishing is allowed, the mode to use, `direct` (default) - fat ping with
+#                          the notification payload specified or `fetch` - the hub fetches the topic URL
+#                          specified in the "publish" request to identify the payload
+public type RemotePublishConfig record {
+    boolean enabled = false;
+    RemotePublishMode mode = PUBLISH_MODE_DIRECT;
```
Comment (pamod) [inline stdlib/websub/src/main/ballerina/websub/commons.bal]: Should we have this as an open record ?

File: stdlib/websub/src/main/ballerina/websub/commons.bal [hunk 6]
```diff
@@ -440,63 +447,59 @@ public type SubscriptionChangeResponse record {
...
-        hubRemotePublishMode = remotePublishMode ?: PUBLISH_MODE_DIRECT;
-    } else {
-        if (remotePublishModeAsConfig.equalsIgnoreCase(REMOTE_PUBLISHING_MODE_FETCH)) {
-            hubRemotePublishMode = PUBLISH_MODE_FETCH;
-        } else if (!remotePublishModeAsConfig.equalsIgnoreCase(REMOTE_PUBLISHING_MODE_DIRECT)) {
-            log:printWarn("unknown publish mode: [" + remotePublishModeAsConfig + "], defaulting to direct mode");
-        }
-    }
-
+                                      default = hubConfiguration["leaseSeconds"] ?: DEFAULT_LEASE_SECONDS_VALUE);
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/commons.bal]: leaseSeconds will never be nil now right? 

Can't we say 
```ballerina
hubLeaseSeconds = config:getAsInt("b7a.websub.hub.leasetime", default = hubConfiguration.leaseSeconds);
```

Comment (MaryamZi) [reply stdlib/websub/src/main/ballerina/websub/commons.bal]: same for hubConfiguration["signatureMethod"] and hubConfiguration["topicRegistrationRequired"]

Comment (chamil321) [reply stdlib/websub/src/main/ballerina/websub/commons.bal]: Need to have nil lifiting there as hubConfiguration is nilable

File: stdlib/websub/src/main/ballerina/websub/commons.bal (hunk 5 above)
Comment (chamil321) [reply stdlib/websub/src/main/ballerina/websub/commons.bal]: +1. Resolved

**Pull request merged**